ecosystem/typescript/sdk_v2/src/transactions/types/typeTag.ts

```diff
@@ -0,0 +1,252 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+/* eslint-disable @typescript-eslint/no-unused-vars */
+/* eslint-disable class-methods-use-this */
+/* eslint-disable max-classes-per-file */
+import { AccountAddress } from "../../core";
+import { Deserializer, Serializable, Serializer } from "../../bcs";
+import { Identifier } from "./identifier";
+import { TypeTagParser } from "./typeTagParser";
+import { TypeTagVariants } from "../../types";
+
+export abstract class TypeTag extends Serializable {
+  abstract serialize(serializer: Serializer): void;
+
+  static deserialize(deserializer: Deserializer): TypeTag {
```

Comment:
Looking at this func, maybe I should do the same to PrivateKey and PublicKey classes - To define the deserialize method at the base level.

Reply:
This abstract factory type of stuff only makes sense when what you're deserializing is represented as an `enum` on the Rust side of things.

In the case of `PrivateKey` and `PublicKey` classes, those are already a part of the `AccountAuthenticator` factory methods, where `AccountAuthenticator` would be the base class responsible for the `static deserialize(...)` method